File: dp/2999-count-the-number-of-powerful-integers.go
package dp

import (
	"strconv"
)

// TODO: Solve it by YOURSELF!!!

// https://leetcode.com/problems/count-the-number-of-powerful-integers/

// Approach: Digital Dynamic Programming
// Time: O(len(finish))
// Space: O(len(finish))
func numberOfPowerfulInt(start int64, finish int64, limit int, s string) int64 {
	var count = func(finish int64) int64 {
		var suff1, suff2 int64 = 0, 0
		var decimal int64 = 1
		for idx := len(s) - 1; idx >= 0; idx-- {
			if finish == 0 {
				return 0
			}
			suff1 += decimal * (finish % 10)
			suff2 += decimal * int64(s[idx]-'0')
			finish /= 10
			decimal *= 10
		}
		var flow = strconv.FormatInt(finish, 10)
		var n = len(flow)
		var dp = make([][2]int64, n+1)
		dp[n][0] = 1
		if suff1 >= suff2 {
			dp[n][1] = 1
		}
		for idx := n - 1; idx >= 0; idx-- {
			dp[idx][0] = int64(limit+1) * dp[idx+1][0]
			if int(flow[idx]-'0') <= limit {
				dp[idx][1] = int64(flow[idx]-'0')*dp[idx+1][0] + dp[idx+1][1]
			} else {
				dp[idx][1] = int64(limit+1) * dp[idx+1][0]
			}
		}
		return dp[0][1]
	}
	return count(finish) - count(start-1)
}

File: dp/2999-count-the-number-of-powerful-integers.py
import unittest

# TODO: Solve it by YOURSELF!!! # pylint: disable=fixme

# https://leetcode.com/problems/count-the-number-of-powerful-integers/

# python3 -m unittest dp/2999-count-the-number-of-powerful-integers.py


class Solution(unittest.TestCase):
    # Approach: Digital Dynamic Programming
    # Time: O(len(finish))
    # Space: O(len(finish))
    def numberOfPowerfulInt(self, start: int, finish: int, limit: int, s: str) -> int:
        def count(finish: int) -> int:
            flow = str(finish)
            n = len(flow)
            preflen = n - len(s)
            if preflen < 0:
                return 0
            dp = [[0] * 2 for _ in range(preflen + 1)]
            dp[preflen][0] = 1
            dp[preflen][1] = 1 if int(flow[preflen:]) >= int(s) else 0
            for idx in range(preflen - 1, -1, -1):
                dp[idx][0] = (limit + 1) * dp[idx + 1][0]
                if int(flow[idx]) <= limit:
                    dp[idx][1] = int(flow[idx]) * dp[idx + 1][0] + dp[idx + 1][1]
                else:
                    dp[idx][1] = (limit + 1) * dp[idx + 1][0]
            return dp[0][1]

        return count(finish) - count(start - 1)

    def test(self):
        for start, finish, limit, s, expected in [
            (1, 6000, 4, "124", 5),
            (15, 215, 6, "10", 2),
            (1000, 2000, 4, "3000", 0),
        ]:
            output = self.numberOfPowerfulInt(start, finish, limit, s)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
